Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp

```diff
@@ -911,139 +965,250 @@ void IntegratePeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
+
+          if (evnt.getSignal() > bg && out[0] < radiusSquared) {
+            // need in V3D for matrix maths later
+            V3D center;
+            for (size_t d = 0; d < nd; ++d) {
+              center[d] = static_cast<double>(center_array[d]);
             }
+            peak_events.emplace_back(center, evnt.getSignal() - bg);
           }
+        }
+      }
+    }
+    box->releaseEvents();
+  } while (MDiter.next());
+  calcCovar(peak_events, pos, radiusSquared, qAxisIsFixed, useCentroid,
+            eigenvects, eigenvals, mean, maxIter);
+}
+
+/**
+ * Calculate the covariance matrix of a spherical region and store the
+ * eigenvectors and eigenvalues that diagonalise the covariance matrix in the
+ * vectors provided
+ *
+ *  @param peak_events    container of center and signal of events in sphere
+ *  @param pos            V3D of nominal peak centre
+ *  @param radiusSquared  radius squared of spherical region for covarariance
+ *  @param qAxisIsFixed   bool to fix an eigenvector along direction pos
+ *  @param useCentroid    bool to use estimated centroid in variance calc
+ *  @param eigenvects     eigenvectors of covariance matrix of spherical region
+ *  @param eigenvals      eigenvalues of covariance matrix of spherical region
+ *  @param mean           container to hold centroid (fixed at pos if not using)
+ *  @param maxIter        max number of iterations in covariance determination
+ */
+void IntegratePeaksMD2::calcCovar(
+    const std::vector<std::pair<V3D, double>> &peak_events, const V3D &pos,
+    const coord_t &radiusSquared, const bool &qAxisIsFixed,
+    const bool &useCentroid, std::vector<V3D> &eigenvects,
+    std::vector<double> &eigenvals, V3D &mean, const int &maxIter) {
+
+  size_t nd = 3;
+
+  // calc threhold mdsq to exclude events over 3 stdevs away
+  boost::math::chi_squared chisq(static_cast<double>(nd));
+  auto mdsq_max = boost::math::quantile(chisq, 0.997);
+  Matrix<double> invCov; // required to calc mdsq
+  double prev_cov_det = DBL_MAX;
+
+  // initialise mean with pos
+  mean = pos;
+  Matrix<double> Pinv(nd, nd);
+  if (qAxisIsFixed) {
+    // transformation from Qlab to Qhat, vhat and uhat,
+    getPinv(pos, Pinv);
+    mean = Pinv * mean;
+  }
+  Matrix<double> cov_mat(nd, nd);
+
+  for (int nIter = 0; nIter < maxIter; nIter++) {
 
+    // reset on each loop
+    cov_mat.zeroMatrix();
+    double w_sum = 0;   // sum of weights
+    size_t nmasked = 0; // num masked events outside 3stdevs
+    auto prev_pos = mean;
+
+    for (size_t ievent = 0; ievent < peak_events.size(); ievent++) {
+
+      const auto event = peak_events[ievent];
+      auto center = event.first;
+      if (qAxisIsFixed) {
+        // transform coords to Q, uhat, vhat basis
+        center = Pinv * center;
+      }
+
+      bool useEvent = true;
+      if (nIter > 0) {
+        // check if point within 3 stdevs of mean (in MD frame)
+        // pos is set equal to mean is useCentroid after each iteration
```

Comment:
I think this should be "...if useCentroid..."?

Reply:
No `nIter > 0` is what I wanted here: before the first iteration we don't have a covariance matrix with which to evaluate mdsq. If `useCentroid == true` then `prev_pos` will always be the nominal position, otherwise it is the centroid from the previous iteration.

Reply:
My comment was about the comment on L1101 but not important - I'll have a look at what you've just pushed

Reply:
Ah sorry - I think I have clarified the comment anyway